compiler-lib/src/sem.rs

```diff
@@ -1,6 +1,163 @@
+
+    // Let's draw some pretty annotations that show we classified classes + their
+    // members and method correctly
+    for ((classsym, membersym, memberytypediscr), decl) in &classes_and_members {
+        // FIXME: Need ClassMember.name be Spanned instead of Symbol
+        let highlightspan = &decl.span;
+        context.diagnostics.info(&Spanned {
+            span: highlightspan.clone(),
+            data: format!("{}.{} ({})", classsym, membersym, memberytypediscr),
+        });
+    }
+
+    // Now the second pass, this one will do (in one traversal)
+    //  - name analysis of the AST
+    //  - type checking of expressions
+    SecondPass::visit(ast, context, &classes_and_members)
+}
+
+struct ClassesAndMembersVisitor;
+
+impl ClassesAndMembersVisitor {
+    pub fn visit<'a, 'f>(
+        ast: &'a ast::AST<'f>,
+        context: &context::Context<'_>,
+    ) -> Result<ClassesAndMembers<'a, 'f>, Error> {
+        let mut res = HashMap::new();
+        Self::do_visit(&NodeKind::from(ast), context, &mut res)?;
+        Ok(res)
+    }
+
+    fn do_visit<'a, 'f>(
+        n: &NodeKind<'a, 'f>,
+        _context: &context::Context<'_>,
+        res: &mut ClassesAndMembers<'a, 'f>,
+    ) -> Result<(), Error> {
+        use self::{ast, NodeKind::*};
+        match n {
+            AST(ast::AST::Program(p)) => {
+                for class in &p.data.classes {
+                    for member in class.members.iter() {
+                        let discr = ast::ClassMemberKindDiscriminants::from(&member.kind);
+                        res.insert(
+                            (Rc::clone(&class.name), Rc::clone(&member.name), discr),
+                            &member,
+                        );
+                    }
+                }
+                Ok(())
+            }
+            x => {
+                return Err(format_err!(
+                    "unexpected ast node passed to first pass: {:?}",
+                    x
+                ))
+            }
+        }
+    }
+}
+
+type ScopeDefs<T> = HashMap<Symbol, T>;
+
+struct Scope<T> {
+    parent: Option<Rc<Scope<T>>>,
+    definitions: ScopeDefs<T>,
+}
+
+impl<T> Scope<T> {
+    pub fn root() -> Rc<Self> {
+        Rc::new(Scope {
+            parent: None,
+            definitions: ScopeDefs::new(),
+        })
+    }
+    pub fn enter_scope(parent: Rc<Self>) -> Self {
+        Scope {
+            parent: Some(parent),
+            definitions: ScopeDefs::new(),
+        }
+    }
+    pub fn leave(self) -> Option<Rc<Self>> {
+        self.parent
+    }
+    pub fn define(&mut self, sym: &Symbol, val: T) -> Result<(), ()> {
+        use std::collections::hash_map::Entry;
+        let key = Rc::clone(sym);
+        match self.definitions.entry(key) {
+            Entry::Occupied(o) => return Err(()),
+            Entry::Vacant(e) => e.insert(val),
+        };
+        Ok(())
+    }
+}
+
+struct SecondPass;
```

Comment:
I would love some comments here. E.g. It would be perfect to list what the "second pass" should do in a doc comment. Same for FirstPass.